File: src/main/java/zielu/gittoolbox/status/BehindStatus.java
package zielu.gittoolbox.status;

import java.util.OptionalInt;

public class BehindStatus {
  private static final BehindStatus EMPTY = new BehindStatus();

  private final int count;
  private final Status status;
  private final Integer delta;

  private BehindStatus(int count, Status status, Integer delta) {
    this.count = count;
    this.status = status;
    this.delta = delta;
  }

  private BehindStatus() {
    this.count = 0;
    this.status = Status.SUCCESS;
    this.delta = null;
  }

  public static BehindStatus create(RevListCount behind) {
    return new BehindStatus(behind.value(), behind.status(), null);
  }

  public static BehindStatus create(RevListCount behind, int delta) {
    return new BehindStatus(behind.value(), behind.status(), delta);
  }

  public static BehindStatus create(int count, int delta) {
    return new BehindStatus(count, Status.SUCCESS, delta);
  }

  public static BehindStatus create(int count) {
    return new BehindStatus(count, Status.SUCCESS, null);
  }

  public static BehindStatus empty() {
    return EMPTY;
  }

  public int behind() {
    return count;
  }

  public Status status() {
    return status;
  }

  public OptionalInt delta() {
    return delta != null ? OptionalInt.of(delta) : OptionalInt.empty();
  }
}

File: src/main/kotlin/zielu/gittoolbox/status/BehindStatus.kt
package zielu.gittoolbox.status

import java.util.OptionalInt

internal data class BehindStatus(
  val behind: Int,
  val status: Status,
  val delta: Int?
) {
  constructor(revListCount: RevListCount) : this(revListCount.value!!, revListCount.status, null)

  constructor(revListCount: RevListCount, delta: Int) : this(revListCount.value!!, revListCount.status, delta)

  constructor(count: Int, delta: Int) : this(count, Status.SUCCESS, delta)

  constructor(count: Int) : this(count, Status.SUCCESS, null)

  @Deprecated("replace with Kotlin syntax", replaceWith = ReplaceWith("delta?."))
  fun delta(): OptionalInt {
    return delta?.let { OptionalInt.of(it) } ?: OptionalInt.empty()
  }

  companion object {
    private val empty = BehindStatus(0)

    @JvmStatic
    fun empty(): BehindStatus = empty
  }
}
